File: dictaite_core/realtime/events.py
"""OpenAI realtime event normalization."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Mapping


class RealtimeEventType(StrEnum):
    SOURCE_DELTA = "source_delta"
    SOURCE_COMPLETED = "source_completed"
    TRANSLATION_DELTA = "translation_delta"
    TRANSLATED_AUDIO_DELTA = "translated_audio_delta"
    SESSION_STATE = "session_state"
    ERROR = "error"
    UNKNOWN = "unknown"


@dataclass(frozen=True, slots=True)
class NormalizedEvent:
    type: RealtimeEventType
    text: str = ""
    item_id: str | None = None
    state: str | None = None
    error: str | None = None


def parse_realtime_event(payload: Mapping[str, Any]) -> NormalizedEvent:
    """Normalize known realtime transcription and translation events."""

    event_type = str(payload.get("type") or "")
    if event_type == "conversation.item.input_audio_transcription.delta":
        return NormalizedEvent(
            RealtimeEventType.SOURCE_DELTA,
            text=_text(payload, "delta"),
            item_id=_optional_str(payload.get("item_id")),
        )
    if event_type == "conversation.item.input_audio_transcription.completed":
        return NormalizedEvent(
            RealtimeEventType.SOURCE_COMPLETED,
            text=_text(payload, "transcript", "text"),
            item_id=_optional_str(payload.get("item_id")),
        )
    if event_type == "session.input_transcript.delta":
        return NormalizedEvent(RealtimeEventType.SOURCE_DELTA, text=_text(payload, "delta"))
    if event_type in {
        "session.output_transcript.delta",
        "response.output_text.delta",
        "response.output_audio_transcript.delta",
    }:
        return NormalizedEvent(RealtimeEventType.TRANSLATION_DELTA, text=_text(payload, "delta"))
    if event_type in {"session.output_audio.delta", "response.audio.delta", "response.output_audio.delta"}:
        return NormalizedEvent(RealtimeEventType.TRANSLATED_AUDIO_DELTA)
    if event_type.startswith("session.") or event_type.startswith("response."):
        return NormalizedEvent(RealtimeEventType.SESSION_STATE, state=event_type)
    if event_type == "error":
        error = payload.get("error")
        if isinstance(error, Mapping):
            message = str(error.get("message") or error.get("code") or "Realtime error")
        else:
            message = str(error or "Realtime error")
        return NormalizedEvent(RealtimeEventType.ERROR, error=message)
    return NormalizedEvent(RealtimeEventType.UNKNOWN, state=event_type or None)


def _text(payload: Mapping[str, Any], *keys: str) -> str:
    for key in keys:
        value = payload.get(key)
        if value is not None:
            return str(value)
    return ""


def _optional_str(value: Any) -> str | None:
    if value in (None, ""):
        return None
    return str(value)

File: src/realtime/events.rs
use serde::Deserialize;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RealtimeEvent {
    SourceDelta {
        item_id: Option<String>,
        text: String,
    },
    SourceCompleted {
        item_id: Option<String>,
        text: String,
    },
    TranslationDelta {
        text: String,
    },
    TranslatedAudioDelta,
    SessionState {
        state: String,
    },
    Error {
        message: String,
    },
    Unknown {
        event_type: Option<String>,
    },
}

#[derive(Deserialize)]
struct RawEvent {
    #[serde(rename = "type")]
    event_type: Option<String>,
    item_id: Option<String>,
    delta: Option<String>,
    transcript: Option<String>,
    text: Option<String>,
    error: Option<serde_json::Value>,
}

pub fn parse_event(value: &serde_json::Value) -> RealtimeEvent {
    let raw: RawEvent = match serde_json::from_value(value.clone()) {
        Ok(raw) => raw,
        Err(_) => return RealtimeEvent::Unknown { event_type: None },
    };
    match raw.event_type.as_deref() {
        Some("conversation.item.input_audio_transcription.delta") => RealtimeEvent::SourceDelta {
            item_id: raw.item_id,
            text: raw.delta.unwrap_or_default(),
        },
        Some("conversation.item.input_audio_transcription.completed") => {
            RealtimeEvent::SourceCompleted {
                item_id: raw.item_id,
                text: raw.transcript.or(raw.text).unwrap_or_default(),
            }
        }
        Some("session.input_transcript.delta") => RealtimeEvent::SourceDelta {
            item_id: None,
            text: raw.delta.unwrap_or_default(),
        },
        Some("session.output_transcript.delta")
        | Some("response.output_text.delta")
        | Some("response.output_audio_transcript.delta") => RealtimeEvent::TranslationDelta {
            text: raw.delta.unwrap_or_default(),
        },
        Some("session.output_audio.delta")
        | Some("response.audio.delta")
        | Some("response.output_audio.delta") => RealtimeEvent::TranslatedAudioDelta,
        Some("error") => RealtimeEvent::Error {
            message: raw
                .error
                .and_then(|error| {
                    error
                        .get("message")
                        .and_then(|message| message.as_str())
                        .map(str::to_string)
                })
                .unwrap_or_else(|| "Realtime error".to_string()),
        },
        Some(other) if other.starts_with("session.") || other.starts_with("response.") => {
            RealtimeEvent::SessionState {
                state: other.to_string(),
            }
        }
        other => RealtimeEvent::Unknown {
            event_type: other.map(str::to_string),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn parses_known_and_unknown_events() {
        let event = parse_event(&json!({
            "type": "conversation.item.input_audio_transcription.delta",
            "item_id": "a",
            "delta": "Hi"
        }));
        assert_eq!(
            event,
            RealtimeEvent::SourceDelta {
                item_id: Some("a".into()),
                text: "Hi".into()
            }
        );
        assert!(matches!(
            parse_event(&json!({"type": "new.event"})),
            RealtimeEvent::Unknown { .. }
        ));
    }

    #[test]
    fn parses_error_and_ignores_translated_audio_delta() {
        let error = parse_event(&json!({
            "type": "error",
            "error": {"message": "bad request"}
        }));
        assert_eq!(
            error,
            RealtimeEvent::Error {
                message: "bad request".into()
            }
        );
        assert_eq!(
            parse_event(&json!({"type": "response.audio.delta", "delta": "secret-audio"})),
            RealtimeEvent::TranslatedAudioDelta
        );
    }

    #[test]
    fn parses_ga_realtime_translation_text_delta() {
        let event = parse_event(&json!({
            "type": "response.output_text.delta",
            "delta": "Hola"
        }));
        assert_eq!(
            event,
            RealtimeEvent::TranslationDelta {
                text: "Hola".into()
            }
        );
    }
}
